File: 2020/08/solution.rb
require_relative '../../lib/advent_helper'
helper = AdventHelper.new(script_root:__dir__)
# input = helper.send(:open_file, 'input.txt').read
# responses = input.split("\n\n").map { |r| r.split(/\s+/).map(&:chars) }
# input = helper.auto_parse
input = helper.line_separated_strings('input.txt')
# input = helper.comma_separated_strings('input.txt')


class AssemblyNode
  attr_reader :input, :operations
  attr_accessor :pos, :acc, :log, :visited
  def initialize(input, options = {})
    @input = input
    @visited = Hash.new
    @log = []
    @operations = options.fetch(:operations, {})
    @pos = 0
    @acc = 0
  end

  def self.run(input, options = {})
    self.new(input, options).run
  end

  def run
    main_loop
    halt
  end

  def main_loop
    return if halt_condition
    logging
    tick
    main_loop
  end

  def tick
    operation = evaluate
    output = operation.call(arguments)
    persist(output)
  end

  def evaluate
    operations[input[pos].split(' ').first]
  end

  def arguments
    {
      pos: pos,
      acc: acc,
      num: input[pos].split(' ').last.to_i
    }
  end

  def persist(output)
    self.acc = output[:acc] if output[:acc]
    self.pos = output[:pos] if output[:pos]
  end

  def logging
    visited[pos] = true
    log << input[pos].dup
  end

  def halt_condition
    visited[pos] || pos >= input.length
  end

  def halt
    acc
  end

  def print_logs
    puts log
  end
end

# Part 1
options = {
  operations: {
    'acc' => lambda { |o| { acc: o[:acc] + o[:num], pos: o[:pos] + 1 } },
    'jmp' => lambda { |o| { pos: o[:pos] + o[:num] } },
    'nop' => lambda { |o| { pos: o[:pos] + 1 } }
  }
}

a = AssemblyNode.run(input, options)

# Part 2
b = nil
(0...input.length).each do |i|
  instrs = input.dup
  op, num = instrs[i].split(' ')
  next if op == 'acc'
  op = op == 'nop' ? 'jmp' : 'nop'
  instrs[i] = [op, num].join(' ')

  node = AssemblyNode.new(instrs, options)
  output = node.run

  if node.pos == instrs.length
    b = output
    break
  end
end



helper.output(a, b)

File: 2020/08/solution.exs
IO.write("\e[H\e[2J")
IO.puts("Running #{__ENV__.file}...")

instructions = __DIR__ <> "/" <> "input.txt"
  |> File.read!()
  |> String.split("\n", trim: true)
  |> Enum.map(fn line -> String.split(line, " ", trim: true) end)
  |> Enum.map(fn [op, num] -> {String.to_atom(op), String.to_integer(num)} end)

defmodule Console do
  def run(instructions, visited \\ %{}, pos \\ 0, acc \\ 0) do
    cond do
      Map.has_key?(visited, pos) || pos == length(instructions) ->
        {acc, pos == length(instructions)}
      true ->
        new_visited = Enum.into(%{pos => true}, visited)
        case Enum.at(instructions, pos) do
          {:acc, number} ->
            run(instructions, new_visited, pos + 1, acc + number)
          {:nop, _} ->
            run(instructions, new_visited, pos + 1, acc)
          {:jmp, number} ->
            run(instructions, new_visited, pos + number, acc)
        end
    end
  end
end

Console.run(instructions) |> elem(0) |> IO.inspect

Enum.each(0..(length(instructions)-1), fn i ->
  {op, number} = Enum.at(instructions, i)
  unless op == :acc do
    modified_instructions = Enum.with_index(instructions) |> Enum.map(fn {instr, index} ->
      if i != index do
        instr
      else
        case {op, number} do
          {:nop, number} -> {:jmp, number}
          {:jmp, number} -> {:nop, number}
        end
      end
    end)
    {acc, valid} = Console.run(modified_instructions)
    if valid do
      IO.puts(acc)
    end
  end
end)
